### PR TITLE
Chore(order): endpoints - Part 1:N

### PR DESCRIPTION
**Pull request by carlos-r-l-rodrigues**
- `archive` and `completeOrder` methods, workflows and endpoint
- Return Reasons endpoints.
- Incomplete skeleton of multiple missing endpoints with TODOs.

### CODE REVIEW

File: packages/core/core-flows/src/order/steps/archive-orders.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
+import { ModuleRegistrationName } from "@medusajs/modules-sdk"
+import { IOrderModuleService } from "@medusajs/types"
+import { StepResponse, createStep } from "@medusajs/workflows-sdk"
+
+type CompleteOrdersStepInput = {
+  order_ids: string[]
```
Comment (olivermrbl) [inline packages/core/core-flows/src/order/steps/archive-orders.ts:6]: nit(non-blocking): camelCase

File: packages/core/core-flows/src/order/steps/complete-orders.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
+import { ModuleRegistrationName } from "@medusajs/modules-sdk"
+import { IOrderModuleService } from "@medusajs/types"
+import { StepResponse, createStep } from "@medusajs/workflows-sdk"
+
+type CompleteOrdersStepInput = {
+  order_ids: string[]
```
Comment (olivermrbl) [inline packages/core/core-flows/src/order/steps/complete-orders.ts:6]: camelCase

File: packages/core/core-flows/src/order/workflows/archive-orders.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
+import { OrderDTO } from "@medusajs/types"
+import { WorkflowData, createWorkflow } from "@medusajs/workflows-sdk"
+import { archiveOrdersStep } from "../steps"
+
+type ArchiveOrdersStepInput = {
+  order_ids: string[]
```
Comment (olivermrbl) [inline packages/core/core-flows/src/order/workflows/archive-orders.ts:6]: ibid.

File: packages/core/core-flows/src/order/workflows/complete-orders.ts [hunk 4]
```diff
@@ -0,0 +1,20 @@
+import { OrderDTO } from "@medusajs/types"
+import { WorkflowData, createWorkflow } from "@medusajs/workflows-sdk"
+import { completeOrdersStep } from "../steps"
+
+type CompleteOrdersStepInput = {
+  order_ids: string[]
```
Comment (olivermrbl) [inline packages/core/core-flows/src/order/workflows/complete-orders.ts:6]: ibid.

File: packages/medusa/src/api-v2/admin/orders/[id]/refund/route.ts [hunk 5]
```diff
@@ -0,0 +1,19 @@
+import { remoteQueryObjectFromString } from "@medusajs/utils"
+import { MedusaRequest, MedusaResponse } from "../../../../../types/routing"
+
+export const GET = async (req: MedusaRequest, res: MedusaResponse) => {
```
Comment (olivermrbl) [inline packages/medusa/src/api-v2/admin/orders/[id]/refund/route.ts:4]: thought(non-blocking): Wonder if we need this? Shouldn't we use /admin/payments/:id/refund instead?

File: packages/medusa/src/api-v2/admin/orders/[id]/cancel/route.ts [hunk 6]
```diff
@@ -0,0 +1,25 @@
+import { remoteQueryObjectFromString } from "@medusajs/utils"
+import {
+  AuthenticatedMedusaRequest,
+  MedusaResponse,
+} from "../../../../../types/routing"
+
+export const GET = async (
```
Comment (olivermrbl) [inline packages/medusa/src/api-v2/admin/orders/[id]/cancel/route.ts:7]: comment for future: Should be `POST`

File: packages/medusa/src/api-v2/admin/orders/[id]/refund/route.ts (hunk 5 above)
Comment (olivermrbl) [inline packages/medusa/src/api-v2/admin/orders/[id]/refund/route.ts:4]: comment for future: Should be `POST`

File: packages/medusa/src/api-v2/admin/orders/fulfillments/[fulfillment_id]/cancel/route.ts [hunk 7]
```diff
@@ -0,0 +1,25 @@
+import { remoteQueryObjectFromString } from "@medusajs/utils"
+import {
+  AuthenticatedMedusaRequest,
+  MedusaResponse,
+} from "../../../../../../types/routing"
+
+export const POST = async (
```
Comment (olivermrbl) [inline packages/medusa/src/api-v2/admin/orders/fulfillments/[fulfillment_id]/cancel/route.ts:7]: thought(non-blocking): Wonder if we need this? We could have `/admin/fulfillments/:id/cancel` instead?

File: packages/medusa/src/api-v2/admin/orders/fulfillments/route.ts [hunk 8]
```diff
@@ -0,0 +1,25 @@
+import { remoteQueryObjectFromString } from "@medusajs/utils"
+import {
+  AuthenticatedMedusaRequest,
+  MedusaResponse,
+} from "../../../../types/routing"
+
+export const POST = async (
```
Comment (olivermrbl) [inline packages/medusa/src/api-v2/admin/orders/fulfillments/route.ts:7]: thought(non-blocking): Wonder if we need this? Why not use `/admin/fulfillments` instead?

File: packages/medusa/src/api-v2/admin/returns/[id]/cancel/route.ts [hunk 9]
```diff
@@ -0,0 +1,35 @@
+import {
+  ContainerRegistrationKeys,
+  remoteQueryObjectFromString,
+} from "@medusajs/utils"
+import {
+  AuthenticatedMedusaRequest,
+  MedusaResponse,
+} from "../../../../../types/routing"
+
+export const GET = async (
```
Comment (olivermrbl) [inline packages/medusa/src/api-v2/admin/returns/[id]/cancel/route.ts:10]: thought: Needs a todo?

File: packages/medusa/src/api-v2/admin/returns/[id]/receive/route.ts [hunk 10]
```diff
@@ -0,0 +1,36 @@
+import {
+  ContainerRegistrationKeys,
+  remoteQueryObjectFromString,
+} from "@medusajs/utils"
+import {
+  AuthenticatedMedusaRequest,
+  MedusaResponse,
+} from "../../../../../types/routing"
+
+export const GET = async (
```
Comment (olivermrbl) [inline packages/medusa/src/api-v2/admin/returns/[id]/receive/route.ts:10]: questions: Needs a todo?

File: packages/medusa/src/api-v2/admin/returns/[id]/route.ts [hunk 11]
```diff
@@ -0,0 +1,35 @@
+import {
+  ContainerRegistrationKeys,
+  remoteQueryObjectFromString,
+} from "@medusajs/utils"
+import {
+  AuthenticatedMedusaRequest,
+  MedusaResponse,
+} from "../../../../types/routing"
+
+export const GET = async (
```
Comment (olivermrbl) [inline packages/medusa/src/api-v2/admin/returns/[id]/route.ts:10]: ibid.

File: packages/medusa/src/api-v2/admin/returns/route.ts [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+import { AdminPostReturnsReqSchemaType } from "./validators"
+
+export const GET = async (
+  req: AuthenticatedMedusaRequest,
+  res: MedusaResponse
+) => {
+  const remoteQuery = req.scope.resolve(ContainerRegistrationKeys.REMOTE_QUERY)
+
+  const queryObject = remoteQueryObjectFromString({
+    entryPoint: "order",
```
Comment (olivermrbl) [inline packages/medusa/src/api-v2/admin/returns/route.ts:19]: question: Should be `return`. Also, do we need a todo here?

File: packages/medusa/src/api-v2/store/return/validators.ts [hunk 13]
```diff
@@ -28,5 +29,12 @@ const ItemSchema = z.object({
 export const StorePostReturnsReqSchema = z.object({
   order_id: z.string(),
   items: z.array(ItemSchema),
-  return_shipping: ReturnShippingSchema.optional(),
+  return_shipping: ReturnShippingSchema,
+  note: z.string().optional(),
+  receive_now: z.boolean().optional(),
+  refund_amount: z.number().optional(),
```
Comment (olivermrbl) [inline packages/medusa/src/api-v2/store/return/validators.ts:35]: todo: You shouldn't be able to pass `refund_amount` from the storefront

File: packages/modules/order/src/services/order-module-service.ts [hunk 14]
```diff
@@ -2577,4 +2579,111 @@ export default class OrderModuleService<
...
+        notAllowed.push(order.id)
+      }
+
+      order.status = OrderStatus.ARCHIVED
+    }
+
+    if (notAllowed.length) {
+      throw new MedusaError(
+        MedusaError.Types.INVALID_DATA,
+        `Pending Order cannot be archived: ${notAllowed.join(", ")}.`
```
Comment (olivermrbl) [inline packages/modules/order/src/services/order-module-service.ts:2625]:
```suggestion
        `Orders ${notAllowed.join(", ")} are completed, canceled, or in draft and cannot be archived`
```

File: packages/modules/order/src/services/order-module-service.ts [hunk 15]
```diff
@@ -2577,4 +2579,111 @@ export default class OrderModuleService<
...
+        notAllowed.push(order.id)
+      }
+
+      order.status = OrderStatus.COMPLETED
+    }
+
+    if (notAllowed.length) {
+      throw new MedusaError(
+        MedusaError.Types.INVALID_DATA,
+        `Canceled Order cannot be completed: ${notAllowed.join(", ")}.`
```
Comment (olivermrbl) [inline packages/modules/order/src/services/order-module-service.ts:2675]:
```suggestion
        `Orders ${notAllowed.join(", ")} are canceled and cannot be completed`
```

**Review by olivermrbl — APPROVED** [review]
Overall LGTM, left a few comments here and there. Still some work left, but this will allows us to finish the UI-side of orders, which is nice.

To not miss these todos, maybe we should create tickets in Linear.

File: packages/medusa/src/api-v2/admin/orders/[id]/refund/route.ts (hunk 5 above)
Comment (carlos-r-l-rodrigues) [reply packages/medusa/src/api-v2/admin/orders/[id]/refund/route.ts:4]: I prefer to use payments directly. Makes more sense to me.
I will remove it.

File: packages/medusa/src/api-v2/admin/orders/fulfillments/route.ts (hunk 8 above)
Comment (carlos-r-l-rodrigues) [reply packages/medusa/src/api-v2/admin/orders/fulfillments/route.ts:7]: You believe the fulfillments workflow could also update necessary things in an Order?

Comment (olivermrbl) [reply packages/medusa/src/api-v2/admin/orders/fulfillments/route.ts:7]: Ah yeah, you are right. There is probably some order-specific steps. Let's keep it 👍

**Comment by carlos-r-l-rodrigues** [comment]
@olivermrbl I am addressing all your comments in a following PR.
Thanks a loto for the review, there are a good chunk of flows left in the Order domain that require us changing a much broader scope than just the Order module, so these details are left behind.

**Pull request merged**